src/media_tools/av_parsers.c

```diff
@@ -1545,6 +1545,595 @@ GF_Err gf_m4a_write_config(GF_M4ADecSpecInfo *cfg, char **dsi, u32 *dsi_size)
+
+	while (frame_size > 8) {
+		u64 obu_size = 0, pos = gf_bs_get_position(bs);
+		ObuType obu_type;
+
+		if (gf_media_aom_av1_parse_obu(bs, &obu_type, &obu_size, state) != GF_OK)
+			return e;
+		assert(obu_size == gf_bs_get_position(bs) - pos);
+
+		av1_populate_state_from_obu(bs, pos, obu_size, obu_type, state);
+
+		frame_size -= obu_size;
+	}
+
+	return e;
+}
+
+typedef enum {
+	KEY_FRAME = 0,
+	INTER_FRAME = 1,
+	INTRA_ONLY_FRAME = 2,
+	SWITCH_FRAME = 3,
+} AV1FrameType;
+
+static void av1_parse_uncompressed_header(GF_BitStream *bs, AV1State *state) {
+	AV1StateFrame *frame_state = &state->frame_state;
+	if (state->frame_id_numbers_present_flag) {
+		//idLen = (additional_frame_id_length_minus_1 + delta_frame_id_length_minus_2 + 3);
+	}
+	//allFrames = (1 << NUM_REF_FRAMES) - 1;
+	if (state->reduced_still_picture_header) {
+		frame_state->key_frame = GF_TRUE;
+	} else {
+		Bool show_existing_frame = gf_bs_read_int(bs, 1);
+		if (show_existing_frame == GF_TRUE) {
+			/*frame_to_show_map_idx	f(3)
+			if (decoder_model_info_present_flag && !equal_picture_interval) {
+				temporal_point_info()
+			}
+			refresh_frame_flags = 0
+			if (frame_id_numbers_present_flag) {
+				display_frame_id	f(idLen)
+			}
+			frame_type = RefFrameType[frame_to_show_map_idx]
+			if (frame_type == KEY_FRAME) {
+				refresh_frame_flags = allFrames
+			}
+			if (film_grain_params_present) {
+				load_grain_params(frame_to_show_map_idx)
+			}*/
+			return;
+		}
+		AV1FrameType frame_type = gf_bs_read_int(bs, 2);
+		frame_state->key_frame = gf_bs_read_int(bs, 1)/*show_frame*/ && frame_type == KEY_FRAME && frame_state->seen_frame_header;
+	}
+}
+
+static void av1_parse_frame_header(GF_BitStream *bs, AV1State *state)
+{
+	AV1StateFrame *frame_state = &state->frame_state;
+	if (frame_state->seen_frame_header == GF_FALSE) {
+		frame_state->seen_frame_header = GF_TRUE;
+		av1_parse_uncompressed_header(bs, state);
+	}
+	/*parsing is voluntarily incomplete*/
+}
+
+GF_Err gf_media_aom_av1_parse_obu(GF_BitStream *bs, ObuType *obu_type, u64 *obu_size, AV1State *state)
+{
+	GF_Err e = GF_OK;
+	Bool obu_has_size_field = GF_FALSE, obu_extension_flag = GF_FALSE;
+	u8 temporal_id = 0, spatial_id = 0;
+	u64 pos = gf_bs_get_position(bs);
+
+	if (!bs || !obu_type || !obu_type || !state)
+		return GF_BAD_PARAM;
+
+	e = av1_parse_obu_header(bs, obu_type, &obu_extension_flag, &obu_has_size_field, &temporal_id, &spatial_id);
+	if (e)
+		return e;
+
+	if (obu_has_size_field) {
+		*obu_size = (u32)leb128(bs, NULL);
+	} else {
+		if (*obu_size >= 1 + obu_extension_flag)
+			*obu_size = *obu_size - 1 - obu_extension_flag;
+		else {
+			GF_LOG(GF_LOG_WARNING, GF_LOG_CODING, ("[AV1] computed OBU size "LLD" (input value = "LLU"). Skipping.\n", *obu_size - 1 - obu_extension_flag, *obu_size));
+			return GF_NON_COMPLIANT_BITSTREAM;
+		}
+	}
+	*obu_size += (gf_bs_get_position(bs) - pos);
+
+	if (*obu_type != OBU_SEQUENCE_HEADER && *obu_type != OBU_TEMPORAL_DELIMITER &&
```

Comment:
This is assuming that only 1 operating point is parsed. If you have 2 operating points you will have to parse twice the bitstream. It might be best to have operating point specific states and to store the OBU there.

Reply:
Please teach me more on this :)